warehouse/ingest-scripts/src/main/resources/bin/util/buildIndexStats.sh

```diff
@@ -0,0 +1,364 @@
+    # when _CountsOnly debug variable is set; check max container size
+
+    # allocate 1/3 memory to mapper io sort buffer but must be less than 2048
+    ((_MapInMemSortBufferMBSize = _MapperMemoryMB / 3))
+    if [[ 1536 -le ${_MapInMemSortBufferMBSize} ]]; then
+        _MapInMemSortBufferMBSize=1536
+    fi
+
+    # enable GC logging if requested - typically for debug purposes
+    # NOTE: there should only be one mapper; same file is used for all mappers
+    if [[ -n "${_GCLog}" || -n "${_GCMapperLog}" ]]; then
+        local -r _logFile="-Xloggc:${_GCLogDir}/gcMap-$$-${_dateStr}.log"
+        local -r _gcLog="${_GCLogOpts} ${_logFile}"
+    fi
+
+    _ChildMapOpts="\
+-Xmx${_MapperMemoryMB}m \
+${_DefaultJVMArgs} \
+${CHILD_INGEST_OPTS} \
+${_HeapDumpOpts} \
+${_gcLog}"
+}
+
+function setReducerJVMParameters() {
+    _ReducerMemoryMB=${_ReducerMemoryMB:-1024}
+
+    # enable GC logging if requested - typically for debug purposes
+    if [[ -n "${_GCLog}" || -n "${_GCMapperLog}" ]]; then
+        local -r _logFile="-Xloggc:${_GCLogDir}/gcReduce-$$-${_dateStr}.log"
+        local -r _gcLog="${_GCLogOpts} ${_logFile}"
+    fi
+
+    _ChildReduceOpts="\
+-Xmx${_ReducerMemoryMB}m \
+${_DefaultJVMArgs} \
+${CHILD_INGEST_OPTS} \
+${_HeapDumpOpts} \
+${_gcLog}"
+}
+
+#==================================
+# Sets required and optional parameters for the map/reduce job.
+function setStatsParameters() {
+    # default values for optional debug parameters
+    # see class files for description
+    #  =======  setting log level for mapper/redcuer/job
+    #  -stats.job.log.level=info
+    #  -stats.mapper.log.level=info
+    #  -stats.reducer.log.level=info
+    #  ======  settings for heartbeat intervals
+    #  -stats.mapper.input.interval=10000000
+    #  -stats.mapper.value.interval=100
+    #  -stats.mapper.uniquecount=false
+    #  -stats.reducer.value.interval=20
+    #  -stats.reducer.counts=false
+    local _StatsDebugOpts=""
+
+    # check for unique counts
+    test -n "${_UniqueCounts}" && {
+        _StatsDebugOpts="${_StatsDebugOpts} -stats.mapper.uniquecount=true -stats.reducer.counts=true"
+    }
+
+    # runtime parameters are set in shard-stats-config.xml
+
+    test -n "${_NumShards}" && {
+        local -r _shardOpts="-num.shards=${_NumShards}"
+    }
+
+    # input and output table names
+    _StatsOpts="
+${_StatsDebugOpts} \
+${_shardOpts} \
+"
+}
+
+#==================================
+# Determines the number of reducers to run based upon the number of shards.
+function calculateReducers() {
+    local -r _reducers=$1
+
+    # divisor for number of reducers based upon the number of shards
+    local -ir _divisor=6
+    local -i _num
+    if [[ -n "${_NumShards}" ]]; then
+        ((_num = _NumShards / _divisor))
+    else
+        IFS=","
+        set -- ${_argDate}
+        unset IFS
+        if [[ $# -gt 1 ]]; then
+            ((_num = _cnt / _divisor))
+        else
+            ((_num = NUM_SHARDS / _divisor))
```

Comment:
NUM_SHARDS is not necessarily the number of shards (see the multi-numshard cached mechanism).  This is probably fine for now but I may create a separate ticket to account for this.,